README.md
# ddot-info

A web app for accessing DDOT bus schedules and real-time data.

This site is built on [Gatsby](https://www.gatsbyjs.com/), pointed at GTFS data stored in Postgres.

You will also need a [Netlify](https://www.netlify.com) account to deploy to production: this project uses Netlify's serverless functions. If you're developing, you can use the Netlify CLI.

## Key features

- bookmarkable route and stop pages
- time aware ui that defaults to show today's service and current trips
- live departure, scheduled departure, and ~~nearby transfer info~~ for every DDOT bus stop
- ~~location aware component to see routes and stops within a 5 minute walk of your current location~~
- digital schedule tables

## Data sources

- [DDOT GTFS](https://data.detroitmi.gov/datasets/ddot-gtfs-file)
- [Clever Devices API](https://myddotbus.com)

# Development

## Installation

`yarn` will install the basic project dependencies.

You'll also want to install `gatsby-cli` and `netlify-cli`.

## Setting up the database

We recommend installing Postgres 12 along with the latest PostGIS extension that works with your version of Postgres.

We use [gtfs-sql-importer](https://github.com/fitnr/gtfs-sql-importer) to import GTFS files into a database.

The database structure is based on the one from **gtfs-sql-importer**, but we add a few helper SQL functions which make new fields and relationships in the GraphQL server, provided by [gatsby-source-pg](https://www.gatsbyjs.com/plugins/gatsby-source-pg/).

You can create a database and import the project database structure from the command line:

```bash
createdb ddotinfo
psql -d ddotinfo -c 'CREATE EXTENSION postgis;'
psql -d ddotinfo < ./gtfs.sql
```

Next, grab the latest version of DDOT's GTFS data and import it to this database using `make` from the **gtfs-sql-importer**:

```bash
curl -o ddot_gtfs.zip "https://detroitmi.gov/Portals/0/docs/deptoftransportation/pdfs/ddot_gtfs.zip"
export PGDATABASE=ddotinfo && make load GTFS=ddot_gtfs.zip
```

You should see the output from SQL insert commands. Importing the DDOT GTFS file typically takes about 2-3 minutes on a newer machine.

### Configuration

Create a `.env.development` file from the given example file, filling in your database details and the API key you've received from DDOT.

### Running the development server

You should be able to run the development server with the `netlify dev` command.

This will run a local Functions server that mirrors how the serverless functions operate in production. In order to deploy this site to production, you would need a Netlify account.

HISTORY.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.1.0 - 2021-04-16

This is still using the same GTFS: 2021-02-08 thru 2021-04-30.

### Added

- Interactive system map at /system-map :tada:
- You can now install ddot.info as a standalone progressive web app, and it will work offline.
- More coherent UI across all pages
- Stop pages now allow you to see where next arrivals are currently
- Site-wide dropdown navigation

### Fixed

- Stops which are endpoints should only show the departure times under "Scheduled departures". This was causing confusion at Fairlane Mall, for example, since it is both the first stop on the northbound trips and the last trip on the southbound. We now longer show the southbound arrival times, only the northbound departure times.
- In rare instances (/stop/60083, Wayne County Community College on Southfield & Outer Dr), a route will serve stops in both directions. In this case we split the times into each direction.

### Changed

- Reclassified 46 Southfield, 80 Villages Direct, 89 Southwest Direct to peak-hour/part-time routes
- Mask alert is now dismissable.

## 1.0.2 - 2021-02-08

Updated to latest GTFS version, valid 2021-02-08 through 2021-04-30.

### Added

- Mask-required alert banner and COVID policy page.

### Fixed

- Weird bug where FontAwesome icons flashed very big on the screen before anything else loaded.
- Google Analytics tracking; we do respect the DNT option.

## 1.0.1 - 2021-01-31

Updated to latest GTFS version, valid 2020-12-21 through 2021-02-28.

A notable change here was that Saturday and Sunday flip-flopped service IDs, which broke some hardcoding.

### Fixed
- footer typo
- error where Grand River directions were N/S instead of E/W
- routes now display on the stop page StopMap again; and more general improvements to StopMap

### Added

- Routes now display on the stop page; click a route to switch which one is displayed on the map.
- Feedback form link on footer.

## 1.0.0 - 2021-01-01


gtfs/routes.py
routes = [
    {
        "id": "67",
        "rt_name": "Cadillac/Harper",
        "color": "#44aa42",
        "rt_id": 6939,
        "timepoints": {
            "eastbound": [8911, 32, 34, 35, 36, 38, 42, 41],
            "westbound": [41, 8963, 44, 46, 47, 49, 51, 8911]
        }
    },

    {
        "id": "68",
        "rt_name": "Chalmers",
        "color": "#9b5ba5",
        "rt_id": 6940,
        "notes": "nonstandard schedule",
        "timepoints": {
            "northbound": [10445, 2075, 9791, 74, 75, 76, 3682],
            "southbound": [3682, 79, 80, 83, 8215, 2076, 10446]
        }
    },

    {
        "id": "52",
        "rt_name": "Chene",
        "color": "#44aa42",
        "rt_id": 6936,
        "timepoints": {
            "northbound": [8892, 87, 118, 89, 90, 91, 92],
            "southbound": [95, 96, 97, 98, 99, 100, 503, 8892]
        }
    },

    {
        "id": "11",
        "rt_name": "Clairmount",
        "color": "#0079c2",
        "rt_id": 6914,
        "timepoints": {
            "eastbound": [103, 2491, 107, 800, 111, 112, 115],
            "westbound": [115, 1236, 118, 8248, 122, 2492, 103]
        }
    },

    {
        "id": "12",
        "rt_name": "Conant",        
        "color": "#9b5ba5",
        "rt_id": 6915,
        "timepoints": {
            "southbound": [127, 7864, 129, 130, 9034, 5525, 132, 10176, 10167],
            "northbound": [10167, 135, 136, 5524, 9035, 138, 139, 140, 127]
        }
    },

    {
        "id": "13",
        "rt_name": "Conner",
        "color": "#9b5ba5",
        "rt_id": 6916,
        "notes": "nonstandard schedule",
        "timepoints": {
            "southbound": [93, 2025, 145, 147, 148, 421],
            "northbound": [421, 153, 154, 10031, 2024, 93]
        }
    },

    {
        "id": "8",
        "rt_name": "Warren",
        "color": "#004445",
        "rt_id": 6911,
        "timepoints": {
            "eastbound": [10187, 7141, 161, 162, 163, 5920, 165, 169, 171, 172, 41],
            "westbound": [41, 175, 48, 177, 181, 5921, 183, 184, 185, 7140, 10187]
        }
    },

    {
        "id": "15",
        "rt_name": "Chicago/Davison",
        "color": "#0079c2",
        "rt_id": 6917,
        "timepoints": {
            "eastbound": [10414, 190, 191, 192, 193, 194, 195, 196, 197, 198],
            "westbound": [198, 200, 201, 202, 203, 204, 205, 206, 207, 10414]
        }
    },

    {
        "id": "16",
        "rt_name": "Dexter",
        "color": "#44aa42",
        "rt_id": 6918,
        "notes": "nonstandard schedule",
        "timepoints": {
            "southbound": [10340, 10325, 212, 213, 214, 215, 216, 218, 220, 221, 8913, 223],
            "northbound": [223, 8913, 224, 225, 227, 229, 230, 231, 6056, 233, 10326, 10340]
        }
    },

    {
        "id": "17",
        "rt_name": "Eight Mile",
        "color": "#0079c2",
        "rt_id": 6919,
        "notes": "nonstandard schedule",
        "timepoints": {
            "eastbound": [236, 238, 10325, 240, 242, 243, 144, 62, 245, 40],
            "westbound": [41, 245, 63, 247, 248, 249, 251, 10326, 253, 236]
        }
    },

    {
        "id": "18",
        "rt_name": "Fenkell",
        "color": "#44aa42",
        "rt_id": 6920,
        "timepoints": {
            "eastbound": [256, 258, 260, 262, 263, 265, 266, 267, 8928],
            "westbound": [8928, 272, 273, 274, 3377, 276, 277, 279, 286, 288]
        }
    },

    {
        "id": "19",
        "rt_name": "Fort",
        "color": "#44aa42",
        "rt_id": 6921,
        "timepoints": {
            "westbound": [8891, 290, 291, 2558, 294, 295],
            "eastbound": [298, 299, 2561, 302, 303, 8891]
        }
    },

    {
        "id": "3",
        "rt_name": "Grand River",
        "color": "#44aa42",
        "rt_id": 6906,
        "notes": "nonstandard schedule",
        "timepoints": {
            "eastbound": [324, 325, 328, 330, 331, 332, 334, 335, 337, 394],
            "westbound": [972, 1102, 346, 347, 2726, 352, 354, 324]
        }
    },

    {
        "id": "10",
        "rt_name": "Greenfield",
        "color": "#004445",
        "notes": "nonstandard schedule",
        "rt_id": 6913,
        "timepoints": {
            "southbound": [210, 10325, 358, 359, 361, 363, 600, 368],
            "northbound": [368, 372, 374, 376, 378, 379, 10326, 210]
        }
    },

    {
        "id": "23",
        "rt_name": "Hamilton",
        "color": "#44aa42",
        "rt_id": 6922,
        "timepoints": {
            "northbound": [8933, 9474, 401, 402, 403, 9658, 10145],
            "southbound": [1045, 9690, 386, 387, 10339, 9475, 8933]
        }
    },

    {
        "id": "9",
        "rt_name": "Jefferson",
        "color": "#004445",
        "rt_id": 6912,
        "timepoints": {
            "eastbound": [9970, 9021, 420, 421, 422, 10178],
            "westbound": [10178, 425, 426, 427, 341, 9970]
        }
    },

    {
        "id": "27",
        "rt_name": "Joy",
        "color": "#44aa42",
        "rt_id": 6923,
        "timepoints": {
            "eastbound": [429, 432, 434, 435, 437, 438, 439, 441, 443, 8910],
            "westbound": [8910, 444, 446, 737, 448, 449, 451, 452, 454, 429]
        }
    },

    {
        "id": "29",
        "rt_name": "Linwood",
        "color": "#44aa42",
        "rt_id": 6924,
        "timepoints": {
            "southbound": [2781, 465, 467, 469, 470, 8912],
            "northbound": [8912, 474, 475, 477, 479, 2781]
        }
    },

    {
        "id": "30",
        "rt_name": "Livernois",
        "color": "#9b5ba5",
        "notes": "nonstandard schedule",
        "rt_id": 6925,
        "timepoints": {
            "southbound": [127, 481, 8987, 482, 483, 484, 486, 487, 488],
            "northbound": [488, 490, 491, 493, 494, 495, 230, 496, 127]
        }
    },

    {
        "id": "31",
        "rt_name": "Mack",
        "color": "#44aa42",
        "rt_id": 6926,
        "timepoints": {
            "westbound": [41, 497, 498, 499, 6334, 502, 1616, 8918],
            "eastbound": [8918, 21, 505, 6336, 508, 509, 510, 41]
        }
    },

    {
        "id": "32",
        "rt_name": "McNichols",
        "color": "#0079c2",
        "notes": "nonstandard schedule",
        "rt_id": 6927,
        "timepoints": {
            "eastbound": [9860, 7230, 514, 515, 517, 519, 524, 510, 41],
            "westbound": [41, 527, 529, 534, 536, 538, 539, 10157, 9860]
        }
    },

    {
        "id": "6",
        "rt_name": "Gratiot",
        "color": "#004445",
        "rt_id": 6909,
        "timepoints": {
            "northbound": [9963, 167, 547, 549, 550, 552, 62],
            "southbound": [554, 555, 550, 558, 178, 179, 9963]
        }
    },

    {
        "id": "2",
        "rt_name": "Michigan",
        "color": "#004445",
        "rt_id": 6905,
        "timepoints": {
            "eastbound": [368, 590, 6578, 591, 1489, 8945],
            "westbound": [8945, 1490, 599, 6577, 600, 368]
        }
    },

    {
        "id": "38",
        "rt_name": "Plymouth",
        "color": "#0079c2",
        "notes": "nonstandard schedule",
        "rt_id": 6928,
        "timepoints": {
            "eastbound": [9776, 9572, 8874, 607, 619, 54, 55, 10449, 59, 613, 1927],
            "westbound": [1927, 614, 66, 10448, 615, 71, 619, 620, 8874, 9549, 9776]
        }
    },

    {
        "id": "39",
        "rt_name": "Puritan",
        "color": "#44aa42",
        "rt_id": 6929,
        "timepoints": {
            "eastbound": [627, 628, 629, 630, 3032, 196],
            "westbound": [201, 634, 635, 636, 637, 638]
        }
    },

    {
        "id": "40",
        "rt_name": "Russell",
        "color": "#44aa42",
        "rt_id": 6930,
        "timepoints": {
            "northbound": [8944, 8777, 7414, 641, 644, 645, 648],
            "southbound": [648, 649, 650, 652, 7415, 655, 8944]
        }
    },

    {
        "id": "41",
        "rt_name": "Schaefer",
        "color": "#9b5ba5",
        "rt_id": 6931,
        "timepoints": {
            "southbound": [656, 657, 659, 660, 6627, 661, 662, 663, 664],
            "northbound": [664, 9561, 662, 668, 6630, 669, 670, 672, 673]
        }
    },

    {
        "id": "42",
        "rt_name": "Mid-City Loop",
        "color": "#d07c32",
        "rt_id": 6932,
        "timepoints": {
            "clockwise": [571, 584, 568, 7831, 4388, 8263, 3381]
        }
    },
     {
        "id": "43",
        "rt_name": "Schoolcraft",
        "color": "#0079c2",
        "rt_id": 6933,
        "notes": "nonstandard schedule",
        "timepoints": {
            "eastbound": [429, 674, 676, 677, 678, 679, 680, 196, 570, 572, 574, 9609, 576],
            "westbound": [7828, 578, 579, 580, 581, 201, 681, 682, 683, 684, 685, 6936, 429]
        }
    },

    {
        "id": "7",
        "rt_name": "Seven Mile",
        "color": "#004445",
        "rt_id": 6910,
        "timepoints": {
            "eastbound": [10157, 689, 6074, 384, 9635, 695, 520, 79, 42, 41],
            "westbound": [41, 8963, 76, 533, 699, 9638, 405, 6073, 705, 10157]
        }
    },

    {
        "id": "46",
        "rt_name": "Southfield",
        "color": "#9b5ba5",
        "rt_id": 6934,
        "timepoints": {
            "southbound": [7282, 707, 708, 709, 710, 712, 843, 368],
            "northbound": [368, 843, 715, 717, 718, 719, 10341, 7282]
        }
    },

    {
        "id": "47",
        "rt_name": "Tireman",
        "color": "#0079c2",
        "notes": "nonstandard schedule",
        "rt_id": 6935,
        "timepoints": {
            "eastbound": [159, 721, 722, 724, 439, 727, 729, 1616],
            "westbound": [1616, 734, 736, 737, 739, 741, 742, 159]
        }
    },

    {
        "id": "5",
        "rt_name": "Van Dyke/Lafayette",
        "color": "#004445",
        "rt_id": 6908,
        "timepoints": {
            "northbound": [8989, 745, 3757, 747, 749, 750, 752, 753, 93],
            "southbound": [93, 754, 755, 757, 758, 760, 3759, 762, 8989]
        }
    },

    {
        "id": "1",
        "rt_name": "Vernor",
        "color": "#004445",
        "rt_id": 6904,
        "timepoints": {
            "eastbound": [6578, 10264, 765, 766, 767, 768, 8915],
            "westbound": [8915, 770, 8774, 772, 773, 10265, 6578]
        }
    },

    {
        "id": "4",
        "rt_name": "Woodward",
        "color": "#004445",
        "rt_id": 6907,
        "timepoints": {
            "southbound": [127, 4604, 785, 787, 788, 1064, 9694],
            "northbound": [9694, 1061, 799, 8848, 802, 4603, 127]
        }
    },

    {
        "id": "54",
        "rt_name": "Wyoming",
        "color": "#9b5ba5",
        "rt_id": 6937,
        "timepoints": {
            "southbound": [127, 250, 805, 807, 808, 809, 810, 811, 322, 323, 815],
            "northbound": [815, 304, 15, 820, 821, 823, 824, 826, 241, 127]
        }
    },

    {
        "id": "60",
        "rt_name": "Evergreen",
        "color": "#9b5ba5",
        "rt_id": 6938,
        "timepoints": {
            "southbound": [10329, 380, 838, 839, 7154, 842, 843, 368],
            "northbound": [368, 843, 844, 7156, 846, 847, 848, 849, 10328]
        }
    },

    {
        "id": "80",
        "rt_name": "Villages Direct",
        "color": "#d07c32",
        "rt_id": 6941,
        "timepoints": {
            "westbound": [10178, 8167, 8148, 4841, 10258, 8263],
            "eastbound": [2507, 10259, 4838, 8147, 8166, 10267]
        }
    },

    {
        "id": "89",
        "rt_name": "Southwest Direct",
        "color": "#d07c32",
        "rt_id": 6942,
        "timepoints": {
            "southbound": [2507, 3412, 1490, 772, 10260, 295],
            "northbound": [295, 10261, 766, 1489, 3411, 8263]
        }
    },

    {
        "id": "92",
        "rt_name": "Rosedale Express",
        "color": "#d07c32",
        "rt_id": 6943,
        "timepoints": {
            "eastbound": [10253, 327, 330, 10231, 8100],
            "westbound": [341, 8441, 351, 847, 7258]
        }
    },

    {
        "id": "95",
        "rt_name": "Ryan Express",
        "color": "#d07c32",
        "rt_id": 6945,
        "timepoints": {
            "northbound": [10204, 10205, 7328, 10207, 10208, 646],
            "southbound": [648, 4931, 10200, 10201, 10202, 10203]
        }
    },

    {
        "id": "96",
        "rt_name": "Joy Express",
        "color": "#d07c32",
        "rt_id": 6946,
        "timepoints": {
            "eastbound": [7158, 7214, 434, 437, 10231, 8100],
            "westbound": [341, 8441, 449, 452, 845, 7151]
        }
    }
]
